### PR TITLE
Use one websocket per browser client

### LINKED ISSUE #2510: Introduce multiplexed grpc-over-websocket proxy to reduce open websockets at runtime
Browsers only support a limited number of websockets at a time (limited by tab/origin/etc), and each grpc stream has a different purpose of some kind - autocomplete, checking for table size changes, and table data subscriptions. The actual value of the limit usually doesn't matter for a simple browser UI, but with PartitionedTable and plot-by implemented, it is necessary to support more concurrent streams than we previously have needed.

The simplest way to support this would be to transmit the "stream id" as part of each frame, so that each end of the connection knows what it is talking about. 

There is an [issue in improbable-eng/grpc-web](https://github.com/improbable-eng/grpc-web/issues/1081) that discusses this, and a [branch](https://github.com/improbable-eng/grpc-web/compare/master...borissmidt:webscocketChannel) with some in-progress work, but the solution there is to wrap with a new protobuf type. I think the simpler approach would be to alter the framing slightly, appending a 4byte stream id ahead of the proper grpc frame, similarly to how the existing websocket proxy uses a "control flow" byte to signal a half-close.

--

Negotiation between client and server to support this multiplexing feature appears to be possible using websocket subprotocols, where a client that supports more than one system could ask the server to pick the "most specific" protocol it can speak. The downside of using this approach would be that each service has to connect directly, rather than letting multiple services share a single socket. This might not matter for us, as we don't have too many services, but it is something to think about. 

Supporting multiple protocols lets us remain compatible with the upstream go proxy and "gracefully" fall back, but with the downside that using the go proxy will go back to the limited stream count behavior.

### PR DESCRIPTION
**Pull request by niloc132**
Creates a new websocket protocol to let the websocket behave only as a transport on which multiple streams, across services, can be conveyed. The websocket will be opened to the path matching the first stream, but after the websocket has started, other streams can be sent on it, using custom metadata/header to give the path to the desired rpc.

By reusing existing paths, the client can signal that it supports either old or new websockets, so it in theory can connect to the old grpc-websockets implementation. Unfortunately, the reference implementation can't handle alternative requested subprotocols, and will just fail the entire request, so this patch breaks our compatibility with that proxy, and so with our netty implementation.

On the other hand, the server will maintain compatibility for both protocols for now, enabling stale js api clients to still be able to connect. We may also find that there are issues with longlived sockets, and might end up wanting to ask the client in some circumstances to use the old implementation.

Fixes #2510

### CODE REVIEW

**Comment by niloc132** [comment]
Note that this is dependent on https://github.com/improbable-eng/grpc-web/pull/1112 (and a subsequent release) to correctly allow clients of a netty server to use this client. Alternatively, we could build our own jetty grpc proxy, or build the JS output differently for netty than jetty so that it defaults to non-multiplexed, or drop support for netty as a server entirely.

**Review by devinrsmith — COMMENTED** [review]
Might be good to have server configuration to disable multiplex, as chatted about.

File: grpc-java/grpc-servlet-websocket-jakarta/src/main/java/io/grpc/servlet/web/websocket/MultiplexedWebsocketStreamImpl.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                        Level.FINEST,
+                        "[{0}] writeFrame: numBytes = {1}, flush = {2}, numMessages = {3}",
+                        new Object[] {logId, frame == null ? 0 : frame.readableBytes(), flush, numMessages});
+            }
+
+            try {
+                if (frame != null) {
+                    int numBytes = frame.readableBytes();
+                    if (numBytes > 0) {
+                        onSendingBytes(numBytes);
```
[the remark this answers is not in the document]
Comment (niloc132) [reply grpc-java/grpc-servlet-websocket-jakarta/src/main/java/io/grpc/servlet/web/websocket/MultiplexedWebsocketStreamImpl.java:97]: No, this is for grpc internally to decide if it should send more. We don't tell it about http2 framing bytes, so we likewise shouldnt tell it about ws framing.

File: grpc-java/grpc-servlet-websocket-jakarta/src/main/java/io/grpc/servlet/web/websocket/MultiplexedWebsocketStreamImpl.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                    }
+
+                    ByteBuffer payload = ByteBuffer.allocate(frame.readableBytes() + 4);
+                    payload.putInt(streamId);
+                    payload.put(((ByteArrayWritableBuffer) frame).bytes, 0, frame.readableBytes());
+                    payload.flip();
+
+                    websocketSession.getBasicRemote().sendBinary(payload);
+                    transportState.runOnTransportThread(() -> transportState.onSentBytes(numBytes));
```
[the remark this answers is not in the document]
Comment (niloc132) [reply grpc-java/grpc-servlet-websocket-jakarta/src/main/java/io/grpc/servlet/web/websocket/MultiplexedWebsocketStreamImpl.java:106]: As above, I don't think that would work.

**Review by devinrsmith — APPROVED** [review]

**Pull request merged**